wrapanapi/systems/openstack.py

```diff
@@ -889,21 +889,7 @@ def volume_exists(self, id):
     def get_volume(self, id):
         return self.capi.volumes.get(id)
 
-    @contextmanager
-    def with_volume(self, *args, **kwargs):
-        """Creates a context manager that creates a single volume with parameters defined via params
-        and destroys it after exiting the context manager
-
-        For arguments description, see the :py:meth:`OpenstackSystem.create_volume`.
-        """
-        volume = self.create_volume(*args, **kwargs)
-        try:
-            yield volume
-        finally:
-            self.delete_volume(volume)
-
-    @contextmanager
-    def with_volumes(self, *configurations, **kwargs):
+    def volume_configurations(self, *configurations, **kwargs):
```

Comment:
There is a use of this context manager in integration_tests. I see your test changes this pattern though.

Why remove it? Why not call create_volume() directly in the test if you don't want to work within a context manager in the test?

Seems like a useful context manager otherwise, and its core function is just create_volume, which can be called otherwise when the context manager isn't needed.

Reply:
Why remove it? It would be a dead code.

The tests were buggy. Once I fixed creating the openstack VMs in the provisioning tests, then I realized the VMs may leak on exception. I probably could fix that by making generalized context managers and use them for deleting vms as well as the volumes:
```
@contextmanager
def cleanuping(the_thing):
    yield the_thing
    the_thing.delete()
```
 But this would make the current ctx manager obsolete.

Anyway, I rather used the callback registration approach  `request.addfinalizer` in the tests. The contextmanagers are quite cumbersome in use when there are more of them. As they require indenting.
```
with cleanuping(create_thing()) as thing1:
    do_some_thing()
    with cleanuping(create_thing2()) as thing2:
        do_some_other_thing()
        with cleanuping(create_thing3()) as thing3:
             do_something(thing1, thing2, thing3
```

Ctx managers may be useful when we need more control about when the destroy should happen (exiting the context), but in tests it is often enough to just register a callback.

Reply:
Sorry, didn't see your response as it was marked resolved without me responding.

Understood, but this is an API change for the library, and while within a test context it may be easy to use the request object to cleanup a created volume or volume configuration, tests are not the only place wrapanapi is used.

If you would like to separate the core functionality of these methods to use their code outside of a context manager, you'll need to maintain the layer of the context manager, and move the core code to newly available methods.  Otherwise we'll need to reach out to consumers of wrapanapi, and evaluate the impact of this API change to RH QE teams.

Reply:
ok